File: ffi/rust/src/keys/tx_key.rs
use crate::ffi::{
  from_tx_key_to_spending_key,
  from_tx_key_to_view_key,
};

crate::macros::impl_key!(TxKey);

impl TxKey {
  pub fn to_spending_key(&self) -> Scalar {
    let blsct_scalar = unsafe { from_tx_key_to_spending_key(self.0.value()) };
    let obj = BlsctObj::from_c_obj(blsct_scalar);
    obj.into()
  }

  pub fn to_view_key(&self) -> Scalar {
    let blsct_scalar = unsafe { from_tx_key_to_view_key(self.0.value()) };
    let obj = BlsctObj::from_c_obj(blsct_scalar);
    obj.into()
  }
}

#[cfg(test)]
mod tests {
  use super::*;
  use crate::{
    initializer::init,
    keys::child_key::ChildKey,
  };

  fn get_tx_key() -> TxKey {
    init();
    let child_key = ChildKey::random().unwrap();
    child_key.to_tx_key()
  }

  #[test]
  fn test_to_spending_key() {
    init();
    let tx_key = get_tx_key();
    tx_key.to_spending_key();
  }

  #[test]
  fn test_to_view_key() {
    init();
    let tx_key = get_tx_key();
    tx_key.to_view_key();
  }
}

File: ffi/ts/src/keys/txKey.ts
import {
  fromTxKeyToSpendingKey,
  fromTxKeyToViewKey
} from '../blsct'

import { Scalar } from '../scalar'

/** Represents a tx key. A tx key is a Scalar and introduces no new functionality; it serves purely as a semantic alias. Both SpendingKey and ViewKey are exclusively derived from a TxKey.
 *
 * Examples:
 * ```ts
 * const { Scalar, ChildKey } = require('navio-blsct')
 * const s = Scalar.random()
 * const ck = new ChildKey(s)
 * const tk = ck.toTxKey()
 * tk.toSpendingKey()
 * tk.toViewKey()
 * ```
 */
export class TxKey extends Scalar {
  constructor(obj?: any) {
    super(obj)
  }

  toSpendingKey(): Scalar {
    const obj = fromTxKeyToSpendingKey(this.value())
    return Scalar.fromObj(obj)
  }

  toViewKey(): Scalar {
    const obj = fromTxKeyToViewKey(this.value())
    return Scalar.fromObj(obj)
  }
}

